docs/RationaleSymbols.md

```diff
@@ -35,32 +35,49 @@ symbol-like data to allow forming non-SSA linkage between disparate elements.
-within an `InnerRefNamespace`.  The `inner_sym` attribute must be an `InnerSymAttr`
-which defines the inner symbols attached to the operation and its fields.
-Operations containing an inner symbol must implement the `InnerSymbol` interface.
+`InnerSymbolTable`.  `InnerSymbolTable` operations can themselves be arbitrarily
+nested, and thus must reside within an `InnerRefNamespace` or `InnerSymbolTable`.
+The `inner_sym` attribute must be an `InnerSymAttr` which defines the inner
+symbols attached to the operation and its fields. Operations containing an inner
+symbol must implement the `InnerSymbol` interface.
 
 Inner Symbols are different from normal symbols due to MLIR symbol table 
 resolution rules.  Specifically, normal symbols are resolved by first going up 
 to the closest parent symbol table and resolving down from there (recursing 
-back down for nested symbol paths).  In FIRRTL and HW, modules define a symbol in a 
-`firrtl.circuit` or `builtin.module` symbol table.  For instances to be able to resolve the 
+back down for nested symbol paths). Conversely, Inner symbol symbol resolution
+is rooted at some encapsulating `InnerRefNamespace`, which defines the top-level
+symbol table hierarchy. In FIRRTL and HW, modules define a symbol in a
+`firrtl.circuit` or `builtin.module` symbol table, which is then interpreted as an
+`InnerRefNamespace`.  For instances to be able to resolve the 
 modules they instantiate, the symbol use in an instance must resolve in the 
-top-level symbol table.  If a module were a symbol table, instances resolving a 
+top-level `InnerRefNamespace`.  If a module were a symbol table, instances resolving a 
 symbol would start from their own module, never seeing other modules (since 
 resolution would start in the parent module of the instance and be unable to go 
-to the global scope).  The second problem arises from nesting.  Symbol 
+to the global scope).  The second problem arises from nesting. MLIR Symbol 
 defining operations must be immediate children of a symbol table.  FIRRTL and HW/SV 
 operations which define an `inner_sym` are grandchildren, at least, of a symbol 
 table and may be much further nested.  Lastly, ports need to define `inner_sym`, 
 something not allowed by normal symbols.
 
+Given that all `InnerSym` references are rooted at some top-level `InnerRefNamespace`,
+inner symbols are not able to be resolved as relative references (i.e. if a
+reference occurs somewhere within a set of nested `InnerSymbolTable` operations,
+and a relative reference is used to reference e.g. a cousin symbol).  
+This design has been chosen to make inner symbol resolution both more robust
+as well as simpler to implement, based on an observation that relative references
+in languages are usually a human convenience, and not necessarily a good IR
+construct. In general, front-ends will generate fully qualified names early in
+the compilation process suitable for use in the IR. And if references truly are
```

Comment:
Specifically, references which are relative to the instance hierarchy are dynamically scoped, which is a different implementation than symbols, which are lexically scoped.

Reply:
Added